src/extensions/customemotes/CustomEmotesLoader.cpp

```diff
@@ -0,0 +1,253 @@
+
+template <>
+inline bool qMapLessThanKey<chatterino::EmoteName>(
+    const chatterino::EmoteName &key1, const chatterino::EmoteName &key2)
+{
+    return key1.string < key2.string;
+}
+
+template <>
+inline bool qMapLessThanKey<chatterino::EmoteNameSet>(
+    const chatterino::EmoteNameSet &key1, const chatterino::EmoteNameSet &key2)
+{
+    return key1.getString() < key2.getString();
+}
+
+namespace chatterino {
+
+namespace {
+
+    Url getEmoteLink(const QJsonObject &urls, const QString &emoteScale)
+    {
+        auto emote = urls.value(emoteScale);
+        if (emote.isUndefined())
+        {
+            return {""};
+        }
+
+        assert(emote.isString());
+
+        return {emote.toString()};
+    }
+    void fillInEmoteData(const QJsonObject &urls, const EmoteName &name,
+                         const QString &tooltip, Emote &emoteData)
+    {
+        auto url1x = getEmoteLink(urls, "1");
+        auto url2x = getEmoteLink(urls, "2");
+        auto url3x = getEmoteLink(urls, "3");
+
+        //, code, tooltip
+        emoteData.name = name;
+        emoteData.images =
+            ImageSet{Image::fromUrl(url1x, 1), Image::fromUrl(url2x, 0.5),
+                     Image::fromUrl(url3x, 0.25)};
+        emoteData.tooltip = {tooltip};
+    }
+}  // namespace
+
+CustomEmotesLoader::CustomEmotesLoader()
+{
+}
+
+void CustomEmotesLoader::loadCustomEmotes()
+{
+    for (const auto &link :
+         getApp()->extensions->customEmotes->items.getVector())
+    {
+        this->load(link);
+    }
+}
+
+void CustomEmotesLoader::loadCustomEmotes(const QUrl &url)
+{
+    this->load(url);
+}
+
+void CustomEmotesLoader::load(const QUrl &url)
+{
+    {
+        std::lock_guard<std::mutex> lock(this->emotesMtx);
+        auto exists = std::find_if(
+            this->emotes.begin(), this->emotes.end(),
+            [&url](const auto &pair) { return pair.first == url; });
+        if (exists != this->emotes.end())
+        {
+            return;
+        }
+        this->emotes.append(QPair(url, nullptr));
+    }
+    NetworkRequest req(url);
+    req.setCaller(QThread::currentThread());
+
+    req.onSuccess([this, url](auto result) -> Outcome {
+        EmotesInfo emoteInfo;
+        emoteInfo.url = url;
+        auto jsonRoot = result.parseJson();
+        auto jsonSets = jsonRoot.value("sets").toArray();
+        auto jsonName = jsonRoot.value("name");
+        emoteInfo.nameExists = !jsonName.isUndefined();
+        if (emoteInfo.nameExists)
+        {
+            emoteInfo.name = jsonName.toString();
+        }
+
+        QString topName =
+            emoteInfo.name.isEmpty() ? "" : (emoteInfo.name + "<br/>");
+
+        for (const auto &subset : jsonSets)
+        {
+            // exists or empty QString
+            auto jsonSubName = subset.toObject().value("name").toString();
+            emoteInfo.subsetNames.push_back(jsonSubName);
+            QString globalTooltip = "<br/>Global Custom Emote<br/>" +
```

Comment:
Wouldn't the api name fit better here?
Like "Global Misaka Emote" and for channels have "Channel Misaka Emote" e. g.
and have emotes sets below it, that way we get 3 lines of text instead of 4

Reply:
That could make the box a bit wide tho